docs/content/sips/000-app-config.md

```diff
@@ -0,0 +1,143 @@
+template = "main"
+date = "2022-03-22T14:53:30Z"
+
+---
+
+Summary: A configuration system for Spin applications.
+
+Owner: lann.martin@fermyon.com
+
+Created: March 22, 2022
+
+Updated: April 1, 2022
+
+## Background
+
+It is common for applications to require configuration at runtime that isn't known at build time or is too sensitive to be stored with build artifacts:
+
+- Logging configuration
+- Per-channel (production, staging, etc) service dependency URLs
+- Database secrets
+
+## Proposal
+
+### Configuration is defined by components and applications
+
+Configuration within a "parent" (component or application) consists of a number of configuration "slots":
+- Slots are uniquely identified within their parent by a string "key"; in order to allow unambiguous conversion to environment variables or file paths, keys are constrained:
+  - Keys must start with a letter (required for env vars)
+  - Keys consist of only lowercase ascii alphanum and `_` (`[a-z0-9_]`)
+    - Only one `_` at a time and not at the end (to allow delimiting in env vars with `__`)
+- A slot must _either_ be marked as "required" _or_ must be given a default value
+- A slot may be marked as "secret", in which case any associated value should be handled with care (e.g. not logged)
+
+```toml
+[config]
+# This simple form...
+key1 = "default_value"
+# ...is equivalent to:
+key1 = { default = "default_value" }
+# required & secret slot 
+key2 = { required = true, secret = true }
+```
+
+Default values can use template strings to reference other slots.
+```toml
+[config]
+key1 = { required = true }
+key2 = "prefix-{{ key1 }}-suffix"
+```
+
+### Components and applications can set configuration of their direct dependencies
+
+In dependency configuration, templates can reference top-level config keys (those in `[config]`), "sibling" keys within the same dependency, and "ancestor" dependant configs.
+
+- Top-level references use just the key name: `{{ top_level_key }}`
+- "Sibling" references use a single `.` prefix: `{{ .sibling_key }}`
+- "Ancestor" references use multiple `.`s: `{{ ..parent_dep_key }}`, `{{ ...grandparent_key }}`
+
+`spin.toml`:
+```toml
+[config]
+app_root = "/app"
+log_file = "{{ app_root }}/log.txt"    # -> "/app/log.txt"
+...
+[[component.config]]
+work_root = "{{ app_root }}/work"      # -> "/app/work"
+work_out = "{{ .work_root }}/output"   # -> "/app/work/output"
+[[component.dependencies.dep1.config]]
+dep_root = "{{ ..work_root }}/dep"     # -> "/app/work/dep"
+```
+
+### Configuration "providers" resolve application configuration
+
+When resolving the value of an application configuration slot, providers will be queried in-order for a value. If no value is returned by any provider, the resolution will either use the default value or fail (if the slot is "required").
+
+Provider configuration is handled by spin at instantiation time (`spin up`).
+
+_Note: Provider configuration is TBD; as TOML it could look like:_
+
+```toml
+[[config-provider]]
```

Comment:
You show this using multiple-table syntax, implying there can be more than one provider.  Maybe worth illustrating why this has value e.g. wanting to keep credentials in a secure vault, but directory names as EVs set by Hippo.

What happens if two providers give values for the same config item?  Error, or first one wins, or last one wins?  Does this mean we need to check every provider for every config item, _just in case_ it has a value?  First or last wins is potentially brittle to someone "tidying" the provider list and not testing well enough 'because it's only tidying up.'

Reply:
Currently:
> providers will be queried in-order for a value

This is an area where I expect us to make changes in the future. Just shooting for a reasonable starting point here.